File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/IndicatorsChart.swift
//
//  IndicatorsChart.swift
//  Swift Examples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

import UIKit

class IndicatorsChart : TKExamplesExampleViewController, UIPopoverControllerDelegate {
    
    var trendlines = [(String, AnyClass)]()
    var indicators = [(String, AnyClass)]()
    let overlayChart = TKChart()
    let indicatorsChart = TKChart()
    let data = StockDataPoint.stockPoints()
    var series: TKChartCandlestickSeries?

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        
        self.trendlines.append(("Simple moving average", TKChartSimpleMovingAverageIndicator.self))
        self.trendlines.append(("Exponential moving average", TKChartExponentialMovingAverageIndicator.self))
        self.trendlines.append(("Weighted moving average", TKChartWeightedMovingAverageIndicator.self))
        self.trendlines.append(("Triangular moving average", TKChartTriangularMovingAverageIndicator.self))
        self.trendlines.append(("Bollinger bands indicator", TKChartBollingerBandIndicator.self))
        self.trendlines.append(("Moving average envelopes", TKChartMovingAverageEnvelopesIndicator.self))
        self.trendlines.append(("Typical price", TKChartTypicalPriceIndicator.self))
        self.trendlines.append(("Weighted close", TKChartWeightedCloseIndicator.self))
        self.trendlines.append(("Median price", TKChartMedianPriceIndicator.self))
        self.trendlines.append(("Modified moving average", TKChartModifiedMovingAverageIndicator.self))
        self.trendlines.append(("Adaptive moving average", TKChartAdaptiveMovingAverageIndicator.self))
        
        self.indicators.append(("Percentage volume oscillator", TKChartPercentageVolumeOscillator.self))
        self.indicators.append(("Percentage price oscillator", TKChartPercentagePriceOscillator.self))
        self.indicators.append(("Absolute volume oscillator", TKChartAbsoluteVolumeOscillator.self))
        self.indicators.append(("MACD indicator", TKChartMACDIndicator.self))
        self.indicators.append(("RSI", TKChartRelativeStrengthIndex.self))
        self.indicators.append(("Accumulation distribution line", TKChartAccumulationDistributionLine.self))
        self.indicators.append(("True range", TKChartTrueRangeIndicator.self))
        self.indicators.append(("Average true range", TKChartAverageTrueRangeIndicator.self))
        self.indicators.append(("Commodity channel index", TKChartCommodityChannelIndex.self))
        self.indicators.append(("Fast stochastic indicator", TKChartFastStochasticIndicator.self))
        self.indicators.append(("Slow stochastic indicator", TKChartSlowStochasticIndicator.self))
        self.indicators.append(("Rate of change", TKChartRateOfChangeIndicator.self))
        self.indicators.append(("TRIX", TKChartTRIXIndicator.self))
        self.indicators.append(("Williams percent", TKChartWilliamsPercentIndicator.self))
        self.indicators.append(("Ease of movement", TKChartEaseOfMovementIndicator.self))
        self.indicators.append(("Detrended price oscillator", TKChartDetrendedPriceOscillator.self))
        self.indicators.append(("Force index", TKChartForceIndexIndicator.self))
        self.indicators.append(("Rapid adaptive variance", TKChartRapidAdaptiveVarianceIndicator.self))
        self.indicators.append(("Standard deviation", TKChartStandardDeviationIndicator.self))
        self.indicators.append(("Relative momentum index", TKChartRelativeMomentumIndex.self))
        self.indicators.append(("On balance volume", TKChartOnBalanceVolumeIndicator.self))
        self.indicators.append(("Price volume trend", TKChartPriceVolumeTrendIndicator.self))
        self.indicators.append(("Positive volume index", TKChartPositiveVolumeIndexIndicator.self))
        self.indicators.append(("Negative volume index", TKChartNegativeVolumeIndexIndicator.self))
        self.indicators.append(("Money flow index", TKChartMoneyFlowIndexIndicator.self))
        self.indicators.append(("Ultimate oscillator", TKChartUltimateOscillator.self))
        self.indicators.append(("Full stochastic indicator", TKChartFullStochasticIndicator.self))
        self.indicators.append(("Market facilitation index", TKChartMarketFacilitationIndex.self))
        self.indicators.append(("Chaikin oscillator", TKChartChaikinOscillator.self))
        
        for (title, _) in self.trendlines {
            self.addOption(title, selector: "addOverlayToChart", inSection: "Trendlines")
        }

        for (title, _) in self.indicators {
            self.addOption(title, selector: "addIndicatorToChart", inSection: "Indicators")
        }
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let exampleBounds = self.view.bounds;
        
        overlayChart.frame = CGRectMake(exampleBounds.origin.x, exampleBounds.origin.y, exampleBounds.size.width, exampleBounds.size.height / 1.5)
        overlayChart.gridStyle.verticalLinesHidden = false
        overlayChart.autoresizingMask = UIViewAutoresizing(rawValue: ~UIViewAutoresizing.None.rawValue)
        self.view.addSubview(overlayChart)
        
        let indicatorsOffset = exampleBounds.origin.y + overlayChart.bounds.size.height + 15
        indicatorsChart.frame = CGRectMake(exampleBounds.origin.x, indicatorsOffset, exampleBounds.size.width, self.view.bounds.size.height - indicatorsOffset)
        indicatorsChart.autoresizingMask = UIViewAutoresizing(rawValue: ~UIViewAutoresizing.None.rawValue)
        self.view.addSubview(indicatorsChart)
        
        series = TKChartCandlestickSeries(items: data)
        
        self.loadCharts()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func didMoveToParentViewController(parent: UIViewController?) {
        super.didMoveToParentViewController(parent)
        if (parent == nil) {
            self.removePanZoomObservers()
        }
    }
    
    func loadCharts() {
        overlayChart.removeAllData()
        indicatorsChart.removeAllData()
        
        let yAxis = TKChartNumericAxis(minimum: 250, andMaximum: 750)
        yAxis.style.labelStyle.textAlignment = TKChartAxisLabelAlignment(rawValue:TKChartAxisLabelAlignment.Right.rawValue | TKChartAxisLabelAlignment.Bottom.rawValue)
        yAxis.style.labelStyle.firstLabelTextAlignment = TKChartAxisLabelAlignment(rawValue:TKChartAxisLabelAlignment.Right.rawValue | TKChartAxisLabelAlignment.Top.rawValue)
        yAxis.allowZoom = true
        series!.yAxis = yAxis
        
        let formatter = NSDateFormatter()
        formatter.dateFormat = "MM/dd/yyyy"
        
        let minDate = formatter.dateFromString("01/01/2011")!
        let maxDate = formatter.dateFromString("01/01/2013")!
        let xAxis = TKChartDateTimeAxis(minimumDate: minDate, andMaximumDate: maxDate)
        xAxis.minorTickIntervalUnit = TKChartDateTimeAxisIntervalUnit.Days
        xAxis.majorTickIntervalUnit = TKChartDateTimeAxisIntervalUnit.Years
        xAxis.majorTickInterval = 1
        xAxis.style.majorTickStyle.ticksHidden = false
        xAxis.style.majorTickStyle.ticksOffset = -3
        xAxis.style.majorTickStyle.ticksWidth = 1.5
        xAxis.style.majorTickStyle.maxTickClippingMode = TKChartAxisClippingMode.Visible
        xAxis.style.majorTickStyle.ticksFill = TKSolidFill(color: UIColor(red: 203.0/255.0, green: 203.0/255.0, blue: 203.0/255.0, alpha: 1.0))
        xAxis.allowZoom = true
        series!.xAxis = xAxis
        
        series!.xAxis!.addObserver(self, forKeyPath: "zoom", options: NSKeyValueObservingOptions.New, context: nil)
        series!.xAxis!.addObserver(self, forKeyPath: "pan", options: NSKeyValueObservingOptions.New, context: nil)
        
        self.addOverlayToChart()
        self.addIndicatorToChart()
    }
    
//MARK: - Events
    
    func addOverlayToChart() {
        let section = self.sections![0] as! TKExamplesSectionInfo
        let indicatorClass = self.trendlines[section.selectedOption].1 as! TKChartFinancialIndicator.Type
        let indicator = indicatorClass.init(series:self.series!)
        indicator.selectionMode = TKChartSeriesSelectionMode.Series

        overlayChart.removeAllData()
        overlayChart.addSeries(self.series!)
        overlayChart.addSeries(indicator)
    }
    
    func addIndicatorToChart() {
        let section = self.sections![1] as! TKExamplesSectionInfo
        let indicatorClass = self.indicators[section.selectedOption].1 as! TKChartFinancialIndicator.Type
        let indicator = indicatorClass.init(series:self.series!)
        indicator.selectionMode = TKChartSeriesSelectionMode.Series

        indicatorsChart.removeAllData()
        indicatorsChart.addSeries(indicator)

        let yAxis = indicatorsChart.yAxis as! TKChartNumericAxis!
        var max = ceil(yAxis.range.maximum!.floatValue)
        var min = floor(yAxis.range.minimum!.floatValue)
        if (max < 0) {
            max *= -1
            min *= -1
        }
        yAxis.range.minimum = min;
        yAxis.range.maximum = max;
        yAxis.majorTickInterval = (max - min) / 2.0;
        yAxis.style.labelStyle.textAlignment = TKChartAxisLabelAlignment(rawValue: TKChartAxisLabelAlignment.Right.rawValue | TKChartAxisLabelAlignment.Bottom.rawValue)
        yAxis.style.lineHidden = false
        
        let xAxis = indicatorsChart.xAxis as! TKChartDateTimeAxis
        xAxis.range = series!.xAxis!.range
        xAxis.style.labelStyle.textHidden = true
        xAxis.zoom = overlayChart.xAxis!.zoom
        xAxis.pan = overlayChart.xAxis!.pan
        xAxis.majorTickIntervalUnit = TKChartDateTimeAxisIntervalUnit.Years
        xAxis.majorTickInterval = 1
    }
    
    func removePanZoomObservers() {
        overlayChart.xAxis!.removeObserver(self, forKeyPath: "zoom")
        overlayChart.xAxis!.removeObserver(self, forKeyPath: "pan")
    }

    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        if keyPath == "zoom" {
            self.indicatorsChart.xAxis!.zoom = overlayChart.xAxis!.zoom
        }
        else {
            indicatorsChart.xAxis!.pan = overlayChart.xAxis!.pan
        }
    }
}

File: Examples/Examples/Docs/ChartTypes/IndicatorsChart.cs
using System;
using System.Drawing;
using System.Collections.Generic;

using Foundation;
using UIKit;
using CoreGraphics;

using TelerikUI;

namespace Examples
{
	[Register("IndicatorsChart")]
	public class IndicatorsChart: XamarinExampleViewController
	{
		TKChart overlayChart;
		TKChart indicatorsChart;
		TKChartSeries series;
		List<StockDataPoint> data;

		public override void ViewDidLoad ()
		{
			this.AddOption ("Simple moving average", "Trendlines", () => AddTrendline (new TKChartSimpleMovingAverageIndicator (this.series)));
			this.AddOption ("Exponential moving average", "Trendlines", () => AddTrendline (new TKChartExponentialMovingAverageIndicator(this.series)));
			this.AddOption ("Weighted moving average", "Trendlines", () => AddTrendline (new TKChartWeightedMovingAverageIndicator(this.series)));
			this.AddOption ("Triangular moving average", "Trendlines", () => AddTrendline (new TKChartTriangularMovingAverageIndicator(this.series)));
			this.AddOption ("Bollinger bands indicator", "Trendlines", () => AddTrendline (new TKChartBollingerBandIndicator(this.series)));
			this.AddOption ("Moving average envelopes", "Trendlines", () => AddTrendline (new TKChartMovingAverageEnvelopesIndicator(this.series)));
			this.AddOption ("Typical price", "Trendlines", () => AddTrendline (new TKChartTypicalPriceIndicator(this.series)));
			this.AddOption ("Weighted close", "Trendlines", () => AddTrendline (new TKChartWeightedCloseIndicator(this.series)));
			this.AddOption ("Median price", "Trendlines", () => AddTrendline (new TKChartMedianPriceIndicator(this.series)));
			this.AddOption ("Modified moving average", "Trendlines", () => AddTrendline (new TKChartModifiedMovingAverageIndicator(this.series)));
			this.AddOption ("Adaptive moving average", "Trendlines", () => AddTrendline (new TKChartAdaptiveMovingAverageIndicator(this.series)));

			this.AddOption ("Percentage volume oscillator", "Indicators", () => AddIndicator (new TKChartPercentageVolumeOscillator (this.series)));
			this.AddOption ("Percentage price oscillator", "Indicators", () => AddIndicator (new TKChartPercentagePriceOscillator(this.series)));
			this.AddOption ("Absolute volume oscillator", "Indicators", () => AddIndicator (new TKChartAbsoluteVolumeOscillator(this.series)));
			this.AddOption ("MACD indicator", "Indicators", () => AddIndicator (new TKChartMACDIndicator(this.series)));
			this.AddOption ("RSI", "Indicators", () => AddIndicator (new TKChartRelativeStrengthIndex(this.series)));
			this.AddOption ("Accumulation distribution line", "Indicators", () => AddIndicator (new TKChartAccumulationDistributionLine(this.series)));
			this.AddOption ("True range", "Indicators", () => AddIndicator (new TKChartTrueRangeIndicator(this.series)));
			this.AddOption ("Average true range", "Indicators", () => AddIndicator (new TKChartAverageTrueRangeIndicator(this.series)));
			this.AddOption ("Commodity channel index", "Indicators", () => AddIndicator (new TKChartCommodityChannelIndex(this.series)));
			this.AddOption ("Fast stochastic indicator", "Indicators", () => AddIndicator (new TKChartFastStochasticIndicator(this.series)));
			this.AddOption ("Slow stochastic indicator", "Indicators", () => AddIndicator (new TKChartSlowStochasticIndicator(this.series)));
			this.AddOption ("Rate of change", "Indicators", () => AddIndicator (new TKChartRateOfChangeIndicator(this.series)));
			this.AddOption ("TRIX", "Indicators", () => AddIndicator (new TKChartTRIXIndicator(this.series)));
			this.AddOption ("Williams percent", "Indicators", () => AddIndicator (new TKChartWilliamsPercentIndicator(this.series)));
			this.AddOption ("Ease of movement", "Indicators", () => AddIndicator (new TKChartEaseOfMovementIndicator(this.series)));
			this.AddOption ("Detrended price oscillator", "Indicators", () => AddIndicator (new TKChartDetrendedPriceOscillator(this.series)));
			this.AddOption ("Force index", "Indicators", () => AddIndicator (new TKChartForceIndexIndicator(this.series)));
			this.AddOption ("Rapid adaptive variance", "Indicators", () => AddIndicator (new TKChartRapidAdaptiveVarianceIndicator(this.series)));
			this.AddOption ("Standard deviation", "Indicators", () => AddIndicator (new TKChartStandardDeviationIndicator(this.series)));
			this.AddOption ("Relative momentum index", "Indicators", () => AddIndicator (new TKChartRelativeMomentumIndex(this.series)));
			this.AddOption ("On balance volume", "Indicators", () => AddIndicator (new TKChartOnBalanceVolumeIndicator(this.series)));
			this.AddOption ("Price volume trend", "Indicators", () => AddIndicator (new TKChartPriceVolumeTrendIndicator(this.series)));
			this.AddOption ("Positive volume index", "Indicators", () => AddIndicator (new TKChartPositiveVolumeIndexIndicator(this.series)));
			this.AddOption ("Negative volume index", "Indicators", () => AddIndicator (new TKChartNegativeVolumeIndexIndicator(this.series)));
			this.AddOption ("Money flow index", "Indicators", () => AddIndicator (new TKChartMoneyFlowIndexIndicator(this.series)));
			this.AddOption ("Ultimate oscillator", "Indicators", () => AddIndicator (new TKChartUltimateOscillator(this.series)));
			this.AddOption ("Full stochastic indicator", "Indicators", () => AddIndicator (new TKChartFullStochasticIndicator(this.series)));
			this.AddOption ("Market facilitation index", "Indicators", () => AddIndicator (new TKChartMarketFacilitationIndex(this.series)));
			this.AddOption ("Chaikin oscillator", "Indicators", () => AddIndicator (new TKChartChaikinOscillator(this.series)));

			base.ViewDidLoad ();

			CGRect exampleBounds = this.View.Bounds;
			CGRect overlayChartBounds = new CGRect (exampleBounds.X, exampleBounds.Y, exampleBounds.Width, exampleBounds.Height/1.5f);
			nfloat indicatorsOffset = exampleBounds.Y + overlayChartBounds.Height + 15;
			CGRect indicatorsChartBounds = new CGRect (exampleBounds.X, indicatorsOffset, exampleBounds.Width, exampleBounds.Height - indicatorsOffset);

			overlayChart = new TKChart (overlayChartBounds);
			overlayChart.GridStyle.VerticalLinesHidden = false;
			overlayChart.AutoresizingMask = ~UIViewAutoresizing.None;
			this.View.AddSubview (overlayChart);

			indicatorsChart = new TKChart (indicatorsChartBounds);
			indicatorsChart.AutoresizingMask = ~UIViewAutoresizing.None;
			this.View.AddSubview (indicatorsChart);

			data = StockDataPoint.LoadStockPoints(-1);
			series = new TKChartCandlestickSeries (data.ToArray ());

			this.loadCharts ();
		}

		void AddTrendline(TKChartFinancialIndicator indicator)
		{
			indicator.SelectionMode = TKChartSeriesSelectionMode.Series;
			overlayChart.RemoveAllData ();
			overlayChart.AddSeries (series);
			overlayChart.AddSeries (indicator);
		}

		void AddIndicator(TKChartFinancialIndicator indicator)
		{
			indicatorsChart.RemoveAllData ();
			indicatorsChart.AddSeries (indicator);

			TKChartNumericAxis yAxis = (TKChartNumericAxis)indicatorsChart.YAxis;

			int max = (int)Math.Ceiling (((NSNumber)yAxis.Range.Maximum).FloatValue);
			int min = (int)Math.Floor (((NSNumber)yAxis.Range.Minimum).FloatValue);
			if (max < 0) {
				max *= -1;
				min *= -1;
			}

			yAxis.Range.Minimum = new NSNumber (min);
			yAxis.Range.Maximum = new NSNumber (max);
			yAxis.MajorTickInterval = new NSNumber ((max - min)/2.0);
			yAxis.Style.LabelStyle.TextAlignment = TKChartAxisLabelAlignment.Right | TKChartAxisLabelAlignment.Bottom;
			yAxis.Style.LineHidden = false;

			TKChartDateTimeAxis xAxis = (TKChartDateTimeAxis)indicatorsChart.XAxis;
			xAxis.Range = series.XAxis.Range;
			xAxis.Style.LabelStyle.TextHidden = true;
			xAxis.Zoom = overlayChart.XAxis.Zoom;
			xAxis.Pan = overlayChart.XAxis.Pan;
			xAxis.MajorTickIntervalUnit = TKChartDateTimeAxisIntervalUnit.Years;
			xAxis.MajorTickInterval = 1;
		}

		void loadCharts()
		{
			TKChartNumericAxis yAxis = new TKChartNumericAxis ();
			yAxis.Range = new TKRange (new NSNumber (250), new NSNumber (750));

			// >> chart-text-align-cs
			yAxis.Style.LabelStyle.TextAlignment = TKChartAxisLabelAlignment.Right | TKChartAxisLabelAlignment.Bottom;
			yAxis.Style.LabelStyle.FirstLabelTextAlignment = TKChartAxisLabelAlignment.Right | TKChartAxisLabelAlignment.Top;
			yAxis.Style.LabelStyle.TextOffset = new UIOffset (0, 0);
			yAxis.Style.LabelStyle.FirstLabelTextOffset = new UIOffset (0, 0);
			// << chart-text-align-cs

			yAxis.AllowZoom = true;
			yAxis.AllowPan = true;
			series.YAxis = yAxis;

			NSDateFormatter formatter = new NSDateFormatter ();
			formatter.DateFormat = "MM/dd/yyyy";

			NSDate minDate = formatter.Parse ("01/01/2011");
			NSDate maxDate = formatter.Parse ("01/01/2013");
			TKChartDateTimeAxis xAxis = new TKChartDateTimeAxis ();
			xAxis.Range = new TKRange (minDate, maxDate);
			xAxis.MinorTickIntervalUnit = TKChartDateTimeAxisIntervalUnit.Days;
			xAxis.MajorTickIntervalUnit = TKChartDateTimeAxisIntervalUnit.Years;
			xAxis.MajorTickInterval = 1;
			xAxis.Style.MajorTickStyle.TicksHidden = false;
			xAxis.Style.MajorTickStyle.TicksOffset = -3;
			xAxis.Style.MajorTickStyle.TicksWidth = 1.5f;
			xAxis.Style.MajorTickStyle.MaxTickClippingMode = TKChartAxisClippingMode.Visible;
			xAxis.Style.MajorTickStyle.TicksFill = new TKSolidFill (UIColor.FromRGB(203/255.0f, 203/255.0f, 203/255.0f));
			xAxis.AllowZoom = true;
			xAxis.AllowPan = true;
			series.XAxis = xAxis;

			AddTrendline (new TKChartSimpleMovingAverageIndicator (this.series));
			AddIndicator (new TKChartPercentageVolumeOscillator (this.series));
		}
	}
}

